### ISSUE TITLE
Deprecate Skopt Suggestion Service

### ISSUE THREAD

**Issue by tenzen-y** [kubeflow/katib#2280]
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Unfortunately, the scikit-optimize was closed on February 29, 2024.

https://github.com/scikit-optimize/scikit-optimize

So we need to stop supporting the skopt suggestion service as well.
I think that we can follow similar deprecation steps as the [MXNet](https://github.com/kubeflow/training-operator/issues/1996).

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

---

Love this feature? Give it a 👍 We prioritize the features with the most 👍

**Comment by tenzen-y** [comment]
@kubeflow/wg-automl-leads WDYT?

**Comment by PeterWrighten** [comment]
May I work for this issue?
/assign

**Comment by tenzen-y** [comment]
/unassign @PeterWrighten 

We need to follow the deprecation flow step by step.

**Comment by PeterWrighten** [comment]
> /unassign @PeterWrighten
> 
> We need to follow the deprecation flow step by step.

Got it. Thanks!

**Comment by andreyvelich** [comment]
That is sad to hear @tenzen-y.

Do we have any alternative to use Bayesian Optimization in Hyperopt or Optuna ?

**Comment by tenzen-y** [comment]
> That is sad to hear @tenzen-y.
> 
> Do we have any alternative to use Bayesian Optimization in Hyperopt or Optuna ?

Unfortunately, both `optuna` and `Hyperopt` don't provide Bayesian sampler.

- https://optuna.readthedocs.io/en/stable/reference/samplers/index.html
- https://hyperopt.github.io/hyperopt/#algorithms

**Comment by tenzen-y** [comment]
If we want to provide Bayesian sample, we need to implement it ourselves.

**Comment by andreyvelich** [comment]
Maybe we could check some Open Source BO implementations.
For example: https://github.com/bayesian-optimization/BayesianOptimization

Also, quite recently @contramundum53 added GPSampler support in Optuna: https://github.com/optuna/optuna/pull/5185

Maybe we can re-use it.
cc @c-bata @a9p for your ideas here.

**Comment by tenzen-y** [comment]
> Also, quite recently @contramundum53 added GPSampler support in Optuna: https://github.com/optuna/optuna/pull/5185

Oh, I didn't find that.

**Comment by tenzen-y** [comment]
So, setting optuna as a default Bayesian sampler might be better.

**Comment by contramundum53** [comment]
@andreyvelich @tenzen-y I'm glad that you are considering integration of Optuna's `GPSampler` into katib!
`GPSampler` is planned to be released in Optuna v3.6, our next minor release. It is still experimental and future breaking changes/bugs are possible (the code is only a month old!), but if it's okay for you, we'll be happy to hear that.

**Comment by tenzen-y** [comment]
> @andreyvelich @tenzen-y I'm glad that you are considering integration of Optuna's `GPSampler` into katib! `GPSampler` is planned to be released in Optuna v3.6, our next minor release. It is still experimental and future breaking changes/bugs are possible (the code is only a month old!), but if it's okay for you, we'll be happy to hear that.

Oh, I see. Thank you for letting us know @contramundum53 
Is there a graduation plan for `GPSampler`? For example, did the optuna community decide when `GPSampler` graduates to stable?

**Comment by contramundum53** [comment]
Sorry no, that will depend on many factors, such as how much the new feature is used, how many feature requests come, how many bugs we encounter, and so on.

**Comment by andreyvelich** [comment]
Thanks for letting us know @contramundum53!

@tenzen-y @johnugeorge What do you think about it ? Should we use `skopt` for 1 Katib/Kubeflow release and add deprecation notice for users who will run experiments with `bayesianoptimization` algorithm ?
Also, implement GPSampler in Optuna Katib Suggestion and explain users how to update Katib Config to use BO from Optuna rather than from skopt ?
After some time we will remove skopt and fully rely on Optuna for `bayesianoptimization` algorithm ?

**Comment by tenzen-y** [comment]
> Sorry no, that will depend on many factors, such as how much the new feature is used, how many feature requests come, how many bugs we encounter, and so on.

@contramundum53 Thank you for letting me know!

**Comment by tenzen-y** [comment]
> @tenzen-y @johnugeorge What do you think about it ? Should we use skopt for 1 Katib/Kubeflow release and add deprecation notice for users who will run experiments with bayesianoptimization algorithm ?
Also, implement GPSampler in Optuna Katib Suggestion and explain users how to update Katib Config to use BO from Optuna rather than from skopt ?
After some time we will remove skopt and fully rely on Optuna for bayesianoptimization algorithm ?

@andreyvelich I totally agree with you.

**Comment by andreyvelich** [comment]
@tenzen-y Should we try to add deprecation notice before the next release ?

**Comment by tenzen-y** [comment]
> @tenzen-y Should we try to add deprecation notice before the next release ?

@andreyvelich Shouldn't be marked as release 0.17, not 0.16?

**Comment by andreyvelich** [comment]
Oh, sorry, my bad

**Comment by tenzen-y** [comment]
> Oh, sorry, my bad

no worries

**Comment by andreyvelich** [comment]
/good-first-issue

If anyone want to work on this issue, we want to migrate to GPSampler from Optuna from skopt to perform bayesian optimization in Katib.

**Comment by akagami-harsh** [comment]
Yes, feel free to take it over. Thanks!
/unassign

**Comment by NarayanaSabari** [comment]
Hi @tenzen-y @andreyvelich,

I'd like to take over this migration task from scikit-optimize to Optuna's GPSampler.

**Background:**
I was investigating issue #2571 (duplicate hyperparameters in Bayesian Optimization), and @andreyvelich correctly pointed out that we should migrate to the maintained Optuna GPSampler first, then test if duplicates still occur with the new algorithm.

**Approach:**
1. Study the previous attempt in PR #2490 to understand what was done
2. Implement the migration from `skopt.Optimizer` to Optuna's `GPSampler`
3. Maintain backward compatibility with existing experiments
4. Add comprehensive tests for categorical, integer, and continuous parameters
5. Update documentation

**Questions:**
- Are there any specific requirements or preferences for the implementation?
- Should this be done incrementally or as a complete replacement?
- Any particular test scenarios you'd like to see covered?

I'll start working on this and keep the issue updated with progress. Let me know if you have any guidance or concerns.

/assign

**Comment by haroon0x** [comment]
@NarayanaSabari are you still working on this ?

**Comment by NarayanaSabari** [comment]
Yes @haroon0x, i am still wait for some review from the maintainers

**Comment by andreyvelich** [comment]
Hi Folks, we are exploring the next iteration of Katib APIs with @aniket2405 and @akshaychitneni: `OptimizationJob`.
This API will implement Kubernetes-native CRD for Hyperparameter Optimization.

It would be valuable to get feedback from the Optuna maintainers, as we’re interested in reusing several algorithms from your framework: 
https://docs.google.com/document/d/1Y8IJ-UdZ7VCEAlax_xEFbbqEi7EB6SfIX4D7ua-xn4M/edit?tab=t.0#heading=h.sva9gnvvmeye


cc @c-bata @contramundum53 @not522
